File: day4.erl
-module(day4).
-import(common, [read_file/1]).
-export([main1/1, main2/1]).

parse_input(Filename) ->
    Lines = read_file(Filename),
    lists:map(
        fun(Line) ->
            [Left, Right] = string:split(Line, ":"),
            [_, IdStr] = string:split(Left, " "),
            {Id, _} = string:to_integer(IdStr),
            [Winners, Numbers] = string:split(Right, "|"),
            {Id, split_numbers(Winners), split_numbers(Numbers)}
        end,
        Lines
    ).

split_numbers(S) ->
    Split = lists:filter(
        fun(Str) -> string:is_empty(Str) == false end,
        string:split(string:trim(S), " ", all)
    ),
    Parsed = lists:map(
        fun(Str) ->
            {N, _} = string:to_integer(Str),
            N
        end,
        Split
    ),
    ordsets:from_list(Parsed).

exp(_, 0) ->
    1;
exp(N, M) ->
    N * exp(N, M - 1).

num_matches({_, Winning, Numbers}) ->
    Matches = lists:filter(fun(N) -> ordsets:is_element(N, Winning) end, Numbers),
    length(Matches).

score_card(C) ->
    NumMatches = num_matches(C),
    if
        NumMatches > 0 ->
            exp(2, NumMatches - 1);
        true ->
            0
    end.

main1(Filename) ->
    Cards = parse_input(Filename),
    Scores = lists:map(fun score_card/1, Cards),
    lists:sum(Scores).

apply_card(Cards, _) when length(Cards) == 0 ->
    [];
apply_card(Cards, Counts) ->
    % io:format("Input ~p ~p ~n", [Cards, Counts]),
    [Card | CardsRest] = Cards,
    [Count | CountsRest] = Counts,
    Matches = num_matches(Card),
    AddedCards = lists:duplicate(Matches, Count),
    NewCounts = lists:map(
        fun({L, R}) -> L + R end,
        lists:zip(AddedCards, CountsRest, {pad, {0, 0}})
    ),
    RecursedCount = apply_card(CardsRest, NewCounts),
    [Count | RecursedCount].

main2(Filename) ->
    Cards = parse_input(Filename),
    Base = lists:map(fun(_) -> 1 end, Cards),
    Counts = apply_card(Cards, Base),
    lists:sum(Counts).

File: day4.exs
defmodule Day4 do
  def read_file(filename) do
    text = File.read!(filename)
    String.split(text, "\n", parts: :infinity, trim: true)
  end

  def split_numbers(s) do
    parts = String.split(s, " ", parts: :infinity, trim: true)
    Enum.map(parts, &String.to_integer/1)
  end

  def parse_line(line) do
    [left, right] = String.split(line, ":", trim: true)
    [_, idString] = String.split(left, " ", trim: true)
    id = String.to_integer(idString)
    [winnersString, numbersString] = String.split(right, "|", trim: true)

    winners = split_numbers(winnersString)
    numbers = split_numbers(numbersString)

    {id, winners, numbers}
  end

  def parse_input(filename) do
    lines = read_file(filename)
    Enum.map(lines, &parse_line/1)
  end

  def exp(a, b) do
    cond do
      b <= 0 ->
        1

      true ->
        a * exp(a, b - 1)
    end
  end

  def score_card(card) do
    # Process.sleep(500)
    {_, winners, numbers} = card
    wset = MapSet.new(winners)
    nset = MapSet.new(numbers)
    inter = MapSet.intersection(wset, nset)
    matches = MapSet.size(inter)

    cond do
      matches == 0 ->
        0

      true ->
        exp(2, matches - 1)
    end
  end

  def main(filename) do
    cards = parse_input(filename)
    tasks = Enum.map(cards, fn card -> Task.async(fn -> score_card(card) end) end)
    scores = Task.await_many(tasks)
    # scores = Enum.map(cards, &score_card/1)
    Enum.sum(scores)
  end
end

IO.inspect(Day4.main("day4.sample"))
